### PR TITLE
Add IAM role and create Glue Crawler task

### PR DESCRIPTION
**Pull request by lalithanagarur**
Creating IAM role to assume for Glue Crawler task
Creating policy to allow access S3 bucket and access kms key to decrypt S3 objects
Attached policies and AWS managed GlueServiceRole policy to role
Creating glue database
Creating glue crawler task

Resolves https://github.com/ministryofjustice/data-platform-support/issues/146

### CODE REVIEW

File: .github/workflows/terraform-aws-analytical-platform-data-production-rds-s3-exports copy.yml
Comment (jacobwoffenden) [inline .github/workflows/terraform-aws-analytical-platform-data-production-rds-s3-exports copy.yml:1]: would imagine you didn't want to commit this?

File: terraform/aws/analytical-platform-data-production/s3-glue-crawler/glue-catalog.tf [hunk 1]
```diff
@@ -0,0 +1,3 @@
+resource "aws_glue_catalog_database" "alpha_vcms_data_glue_catalog" {
```
Comment (jacobwoffenden) [inline terraform/aws/analytical-platform-data-production/s3-glue-crawler/glue-catalog.tf]:
```suggestion
resource "aws_glue_catalog_database" "alpha_vcms_data" {
```
Don't need to specify resource type in name

File: terraform/aws/analytical-platform-data-production/s3-glue-crawler/glue-crawler.tf [hunk 2]
```diff
@@ -0,0 +1,9 @@
+resource "aws_glue_crawler" "vmcs_data_crawler" {
+  database_name = aws_glue_catalog_database.alpha_vcms_data_glue_catalog.name
+  name          = "vcms-data-crawler"
```
Comment (jacobwoffenden) [inline terraform/aws/analytical-platform-data-production/s3-glue-crawler/glue-crawler.tf]: update name and resource name to be use alpha prefix, drop `crawler` from resource name
```suggestion
resource "aws_glue_crawler" "alpha_vmcs_data" {
  database_name = aws_glue_catalog_database.alpha_vcms_data_glue_catalog.name
  name          = "alpha-vcms-data"
```

File: terraform/aws/analytical-platform-data-production/s3-glue-crawler/iam-policies.tf [hunk 3]
```diff
@@ -0,0 +1,26 @@
+data "aws_iam_policy_document" "crawler_policy" {
```
Comment (jacobwoffenden) [inline terraform/aws/analytical-platform-data-production/s3-glue-crawler/iam-policies.tf]:
```suggestion
data "aws_iam_policy_document" "alpha_vcms_data_crawler_policy" {
```

this needs scoping, incase we want to deploy more crawlers

File: terraform/aws/analytical-platform-data-production/s3-glue-crawler/iam-policies.tf [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+  statement {
+    sid    = "AllowS3"
+    effect = "Allow"
+    actions = [
+      "s3:GetObject",
+      "s3:PutObject"
+    ]
+    resources = [
+      "arn:aws:s3:::alpha-vcms-data/*"
+    ]
```
Comment (jacobwoffenden) [inline terraform/aws/analytical-platform-data-production/s3-glue-crawler/iam-policies.tf]:
```suggestion
    resources = ["arn:aws:s3:::alpha-vcms-data/*"]
```

File: terraform/aws/analytical-platform-data-production/s3-glue-crawler/iam-policies.tf [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+  }
+  statement {
+    sid    = "AllowKMS"
+    effect = "Allow"
+    actions = [
+      "kms:Decrypt"
+    ]
```
Comment (jacobwoffenden) [inline terraform/aws/analytical-platform-data-production/s3-glue-crawler/iam-policies.tf]:
```suggestion
    actions = ["kms:Decrypt"]
```

File: terraform/aws/analytical-platform-data-production/s3-glue-crawler/iam-policies.tf [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+    resources = [data.aws_kms_key.by_alias.arn]
+  }
+}
+
+resource "aws_iam_policy" "vcms_crawler_policy" {
+  name   = "vcms-crawler-policy"
```
Comment (jacobwoffenden) [inline terraform/aws/analytical-platform-data-production/s3-glue-crawler/iam-policies.tf]:
```suggestion
resource "aws_iam_policy" "alpha_vcms_data_crawler" {
  name   = "alpha-vcms-data-crawler"
```

File: terraform/aws/analytical-platform-data-production/s3-glue-crawler/iam-role-policy-attachments.tf [hunk 7]
```diff
@@ -0,0 +1,9 @@
+resource "aws_iam_role_policy_attachment" "vcms_crawler_policy" {
```
Comment (jacobwoffenden) [inline terraform/aws/analytical-platform-data-production/s3-glue-crawler/iam-role-policy-attachments.tf]:
```suggestion
resource "aws_iam_role_policy_attachment" "alpha_vcms_data_crawler" {
```

File: terraform/aws/analytical-platform-data-production/s3-glue-crawler/iam-roles.tf [hunk 8]
```diff
@@ -0,0 +1,16 @@
+data "aws_iam_policy_document" "glue_policy" {
```
Comment (jacobwoffenden) [inline terraform/aws/analytical-platform-data-production/s3-glue-crawler/iam-roles.tf]:
```suggestion
data "aws_iam_policy_document" "glue_assume_role_policy" {
```

File: terraform/aws/analytical-platform-data-production/s3-glue-crawler/iam-roles.tf [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+    actions = ["sts:AssumeRole"]
+    principals {
+      type        = "Service"
+      identifiers = ["glue.amazonaws.com"]
+    }
+  }
+}
+
+resource "aws_iam_role" "glue_vcms_crawler_role" {
+  name               = "AWSGlueServiceRole-vcms-crawler"
```
Comment (jacobwoffenden) [inline terraform/aws/analytical-platform-data-production/s3-glue-crawler/iam-roles.tf]:
```suggestion
resource "aws_iam_role" "alpha_vcms_data_crawler" {
  name               = "alpha-vcms-data-crawler"
```

File: terraform/aws/analytical-platform-data-production/s3-glue-crawler/kms-keys.tf [hunk 10]
```diff
@@ -0,0 +1,3 @@
+data "aws_kms_key" "by_alias" {
+  key_id = "alias/rds-s3-export"
+}
```
Comment (jacobwoffenden) [inline terraform/aws/analytical-platform-data-production/s3-glue-crawler/kms-keys.tf]: move to data.tf

**Review by jacobwoffenden — CHANGES_REQUESTED** [review]

File: .github/workflows/terraform-aws-analytical-platform-data-production-rds-s3-exports copy.yml
Comment (lalithanagarur) [reply .github/workflows/terraform-aws-analytical-platform-data-production-rds-s3-exports copy.yml:1]: Yep, removed the 'copy' from the name.

File: terraform/aws/analytical-platform-data-production/s3-glue-crawler/iam-policies.tf [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+  statement {
+    sid       = "AllowKMS"
+    effect    = "Allow"
+    actions   = ["kms:Decrypt"]
+    resources = [data.aws_kms_key.by_alias.arn]
+  }
+}
+
+resource "aws_iam_policy" "alpha_vcms_crawler_policy" {
+  name   = "alpha-vcms-crawler-policy"
```
Comment (jacobwoffenden) [inline terraform/aws/analytical-platform-data-production/s3-glue-crawler/iam-policies.tf]:
```suggestion
  name   = "alpha-vcms-crawler"
```

File: terraform/aws/analytical-platform-data-production/s3-glue-crawler/iam-roles.tf [hunk 12]
```diff
@@ -0,0 +1,16 @@
...
+      type        = "Service"
+      identifiers = ["glue.amazonaws.com"]
+    }
+  }
+}
+
+resource "aws_iam_role" "alpha_vcms_data_crawler" {
+  name               = "alpha-vcms-data-crawler"
+  assume_role_policy = data.aws_iam_policy_document.glue_assume_policy.json
+}
```
Comment (jacobwoffenden) [inline terraform/aws/analytical-platform-data-production/s3-glue-crawler/iam-roles.tf]:
```suggestion
resource "aws_iam_role" "alpha_vcms_data_crawler" {
  name                = "alpha-vcms-data-crawler"
  assume_role_policy  = data.aws_iam_policy_document.glue_assume_policy.json
  managed_policy_arns = [
    aws_iam_policy.alpha_vcms_crawler_policy.arn,
    "arn:aws:iam::aws:policy/AWSGlueServiceRole"
  ]
}
```

File: terraform/aws/analytical-platform-data-production/s3-glue-crawler/iam-role-policy-attachments.tf
Comment (jacobwoffenden) [inline terraform/aws/analytical-platform-data-production/s3-glue-crawler/iam-role-policy-attachments.tf:1]: Can these in favour of the `managed_policy_arns` suggestion on `aws_iam_role.alpha_vcms_data_crawler`

**Review by jacobwoffenden — CHANGES_REQUESTED** [review]

**Review by jacobwoffenden — APPROVED** [review]
approved, but I'll need another @ministryofjustice/data-platform-core-infra engineer to approve

**Review by julialawrence — APPROVED** [review]

**Pull request merged**